File: IViewer/UI/Options.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Windows.Forms;

namespace Cselian.IViewer.UI
{
	public partial class Options : Form
	{
		public Options(List<ListContext> history = null)
		{
			InitializeComponent();
			Load += Options_Load;
			SCKey.KeyDown += SCKey_KeyDown;
			MenuItems.DoubleClick += MenuItems_DoubleClick;
			KnownFolders.KeyDown += KnownFolders_KeyDown;
			KnownFolders.AfterLabelEdit += KnownFolders_AfterLabelEdit;
			Save.Click += Save_Click;

			if (history != null)
			{
				History.DataSource = history;
				LoadFaves();
				TabOptions.TabPages.Remove(TabMenuShortcuts);
				TabOptions.TabPages.Remove(TabFolders);
				Faves.SelectedNode = Faves.Nodes[0];
			}
			else
			{
				TabOptions.TabPages.Remove(TabFavorites);
			}
		}

		public bool SavedKnownFolders { get; set; }

		public bool SavedFavorites { get; set; }

		private void Options_Load(object sender, System.EventArgs e)
		{
			if (TabOptions.TabPages.Contains(TabMenuShortcuts))
				InitMenu();
			if (TabOptions.TabPages.Contains(TabFolders))
				InitKnownFolders();
		}

		private void Save_Click(object sender, System.EventArgs e)
		{
			if (TabOptions.TabPages.Contains(TabFolders))
				SaveKnownFolders();
			if (TabOptions.TabPages.Contains(TabFavorites))
				SaveFaves();

			DialogResult = System.Windows.Forms.DialogResult.OK;
			Close();
		}

		#region "Menu"

		private void InitMenu()
		{
			return; // TODO: Fix this
			MenuItems.Items.Clear();
			foreach (var mnu in Main.Menu.AllItems)
			{
				var name = mnu.Name.Substring(3);

				foreach (var mnuItm in mnu.DropDownItems.OfType<ToolStripMenuItem>())
				{
					if (mnuItm.Name.StartsWith(name))
					{
						var itm = MenuItems.Items.Add(new ListViewItem(mnuItm.Text.Replace("&", string.Empty)));
						itm.SubItems.Add(name);
						itm.SubItems.Add(GetSCString(mnuItm.ShortcutKeys));
					}
				}
			}
		}

		private string GetSCString(Keys kys)
		{
			var result = kys.ToString();
			if (result.IndexOf(", ") != -1)
			{
				var rev = result.Split(',');
				result = string.Empty;
				for (int i = rev.Length; i >= 0; i += -1)
				{
					result += rev[i].Trim();
					if (i != 0)
					{
						result += " + ";
					}
				}
			}

			return result;
		}

		private void SCKey_KeyDown(object sender, KeyEventArgs e)
		{
			if (e.KeyCode != Keys.Return & e.KeyCode != Keys.ControlKey & e.KeyCode != Keys.Alt & e.KeyCode != Keys.ShiftKey)
			{
				Keys key = default(Keys);

				if (e.Control)
				{
					key = key | Keys.ControlKey;
				}

				if (e.Alt)
				{
					key = key | Keys.Alt;
				}

				if (e.Shift)
				{
					key = key | Keys.ShiftKey;
				}

				key = key | e.KeyCode;

				SCKey.Text = GetSCString(e.KeyCode);
			}

			e.SuppressKeyPress = true;
		}

		private void MenuItems_DoubleClick(object sender, System.EventArgs e)
		{
			SCKey.Visible = true;
			SCKey.Top = MenuItems.Top + MenuItems.SelectedItems[0].Position.Y;
			SCKey.Text = MenuItems.SelectedItems[0].SubItems[2].Text;
			SCKey.Focus();
			SCKey.SelectAll();
		}

		#endregion

		#region KnownFolders

		private void InitKnownFolders()
		{
			KnownFolders.Items.Clear();
			foreach (var path in VidEngine.Settings.KnownFolders)
			{
				var itm = KnownFolders.Items.Add(path);
				itm.SubItems.Add(Directory.Exists(path).ToString());
			}
		}

		private void KnownFolders_KeyDown(object sender, KeyEventArgs e)
		{
			if (e.KeyCode == Keys.F2)
			{
				KnownFolders.SelectedItems[0].BeginEdit();
			}
			else if (e.KeyCode == Keys.Insert)
			{
				var itm = new ListViewItem("New Known Folder");
				itm.SubItems.Add(string.Empty);
				KnownFolders.Items.Add(itm);
				itm.Selected = true;
				itm.BeginEdit();
			}
			else if (e.KeyCode == Keys.Delete)
			{
				KnownFolders.Items.Remove(KnownFolders.SelectedItems[0]);
			}
		}

		private void KnownFolders_AfterLabelEdit(object sender, LabelEditEventArgs e)
		{
			KnownFolders.Items[e.Item].SubItems[1].Text = Directory.Exists(e.Label).ToString();
		}

		private void SaveKnownFolders()
		{
			SavedKnownFolders = true;
			var fols = new List<string>();
			foreach (ListViewItem itm in KnownFolders.Items)
			{
				fols.Add(itm.Text);
			}

			VidEngine.Settings.KnownFolders = fols.ToArray();
			VidEngine.SaveSettings();
			VidEngine.MainInst.ReloadKnownFols();
		}

		#endregion

		#region History / Favorite List Views

		private void History_MouseDoubleClick(object sender, MouseEventArgs e)
		{
			AddChild((ListContext)History.SelectedItem);
		}

		private void History_KeyDown(object sender, KeyEventArgs e)
		{
			if (e.KeyCode == Keys.Delete)
			{
				if (History.SelectedItem != null)
				{
					var itm = (ListContext)History.SelectedItem;
					var list = (List<ListContext>)History.DataSource;
					var ix = History.SelectedIndex;
					list.Remove(itm);
					History.SelectedItem = list.IndexOrDefault(ix);
				}
			}
		}

		private void Faves_KeyDown(object sender, KeyEventArgs e)
		{
			if (e.KeyCode == Keys.Insert)
			{
				AddChild(new ListContext(ListContext.Types.Group));
			}
			else if (e.KeyCode == Keys.Delete)
			{
				var node = Faves.SelectedNode;
				if (node.Parent == null) return;
				var parent = (ListContext)node.Parent.Tag;
				var child = (ListContext)node.Tag;
				parent.Children.Remove(child);
				node.Parent.Nodes.Remove(node);
			}
		}

		private void AddChild(ListContext ctx)
		{
			var node = Faves.SelectedNode;
			var parent = (ListContext)node.Tag;

			parent.Children.Add(ctx);
			var sub = node.Nodes.Add(ctx.ToString());
			node.ExpandAll();
			sub.Tag = ctx;
		}

		private void LoadFaves()
		{
			var faves = IOHelper.Load<ListContext>("Faves.xml");
			if (faves == null)
				Faves.Nodes[0].Tag = new ListContext(ListContext.Types.Group);
			else
				AddFaveNodes(Faves.Nodes[0], faves);

			Faves.Nodes[0].ExpandAll();
		}

		public static void AddFaveNodes(TreeNode parent, ListContext match)
		{
			parent.Tag = match;
			foreach (var item in match.Children)
			{
				var node = parent.Nodes.Add(item.ToString());
				AddFaveNodes(node, item);
			}
		}

		private void SaveFaves()
		{
			SavedFavorites = true; 
			var node = (ListContext)Faves.Nodes[0].Tag;
			IOHelper.Save(node, "Faves.xml");
		}

		#endregion
	}
}

File: IViewer/VB/Options.vb
Imports System.Windows.Forms.Design
Imports System.IO

Public Class Options

	Public opener As Vid
	Private dirtySettings As List(Of AppConstants)

	Public Sub New(ByVal owner As Vid)
		opener = owner
		' This call is required by the Windows Form Designer.
		InitializeComponent()

		' Add any initialization after the InitializeComponent() call.
	End Sub

	Private Sub Options_Load(ByVal sender As Object, ByVal e As System.EventArgs) Handles Me.Load
		InitMenu()
		InitFavorites()
	End Sub

	Private Sub InitFavorites()
		Favorites.Items.Clear()
		For Each path As String In AppSettings.Favorites
			Favorites.Items.Add(path)
			Favorites.Items(Favorites.Items.Count - 1).SubItems.Add(IO.Directory.Exists(path))
		Next
	End Sub

	Private Sub SaveSettings()

	End Sub

#Region "Menu"

	Private Sub InitMenu()
		MenuItems.Items.Clear()
		For Each mnu As ToolStripMenuItem In opener.VidMenu.Items
			Dim name As String = mnu.Name.Substring(3)
			'Dim nameLen As Integer = name.Length

			For Each obj As Object In mnu.DropDownItems
				Dim mnuItm As ToolStripMenuItem = TryCast(obj, ToolStripMenuItem)
				If mnuItm IsNot Nothing AndAlso mnuItm.Name.StartsWith(name) Then
					Dim itm As ListViewItem = MenuItems.Items.Add(New ListViewItem(mnuItm.Text.Replace("&", "")))
					'mnuItm.Name.Substring(nameLen, mnuItm.Name.LastIndexOf("Mnu") - nameLen)))
					itm.SubItems.Add(name)
					itm.SubItems.Add(GetSCString(mnuItm.ShortcutKeys))
				End If
			Next
		Next
	End Sub

	Private Function GetSCString(ByVal kys As Keys) As String
		Dim result As String = ""
		result = kys.ToString
		If result.IndexOf(", ") <> -1 Then
			Dim rev() As String = result.Split(", ")
			result = ""
			For i As Integer = UBound(rev) To LBound(rev) Step -1
				result += rev(i).Trim()
				If i <> 0 Then result += " + "
			Next
		End If
		Return result
	End Function

	Private Sub SCKey_KeyDown(ByVal sender As Object, ByVal e As System.Windows.Forms.KeyEventArgs) Handles SCKey.KeyDown
		If e.KeyCode <> Keys.Return And e.KeyCode <> Keys.ControlKey And e.KeyCode <> Keys.Alt And e.KeyCode <> Keys.ShiftKey Then
			Dim key As Keys
			'SCKey.Text = ""
			If e.Control Then key = key Or Keys.ControlKey ' SCKey.Text = "Control + "
			If e.Alt Then key = key Or Keys.Alt 'SCKey.Text += "Alt + "
			If e.Shift Then key = key Or Keys.ShiftKey 'SCKey.Text += "Shift + "
			key = key Or e.KeyCode
			SCKey.Text = GetSCString(e.KeyCode)
		End If
		e.SuppressKeyPress = True
	End Sub

	Private Sub MenuItems_DoubleClick(ByVal sender As Object, ByVal e As System.EventArgs) Handles MenuItems.DoubleClick
		SCKey.Visible = True
		SCKey.Top = MenuItems.Top + MenuItems.SelectedItems(0).Position.Y
		SCKey.Text = MenuItems.SelectedItems(0).SubItems(2).Text
		SCKey.Focus()
		SCKey.SelectAll()
	End Sub

#End Region

	Private Sub Favorites_KeyDown(ByVal sender As System.Object, ByVal e As System.Windows.Forms.KeyEventArgs) Handles Favorites.KeyDown
		If e.KeyCode = Keys.F2 Then
			Favorites.SelectedItems(0).BeginEdit()
		ElseIf e.KeyCode = Keys.Insert Then
			Dim itm As New ListViewItem("New Favorite Folder")
			itm.SubItems.Add("")
			Favorites.Items.Add(itm)
			itm.Selected = True
			itm.BeginEdit()
		ElseIf e.KeyCode = Keys.Delete Then
			Favorites.Items.Remove(Favorites.SelectedItems(0))
		End If
	End Sub

	Private Sub Favorites_AfterLabelEdit(ByVal sender As System.Object, ByVal e As System.Windows.Forms.LabelEditEventArgs) Handles Favorites.AfterLabelEdit
		Favorites.Items(e.Item).SubItems(1).Text = Directory.Exists(e.Label).ToString
	End Sub

	Private Sub Save_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles Save.Click
		Dim fols As New List(Of String)
		For Each itm As ListViewItem In Favorites.Items
			fols.Add(itm.Text)
		Next
		AppSettings.Favorites = fols.ToArray
		Vid.ReloadFavorites()
	End Sub
End Class